File: Mini-Instagram/Friend.cpp
#include "Friend.h"

Friend::Friend() {}

Friend::~Friend() {}

bool Friend::haveActiveRelation(const string& user1, const string& user2) {
	ListNode<GraphNode<string>>* user1Node{ friends.findNode(user1) };
	ListNode<Edge<string>>* user1RelationUser2{ friends.findEdge(user1, user2) };

	if (user1Node == nullptr || user1RelationUser2 == nullptr || user1RelationUser2->data.weight != "active") {
		return false;
	}

	return true;
}

bool Friend::havePendingRelation(const string& user1, const string& user2) {
	ListNode<GraphNode<string>>* user1Node{ friends.findNode(user1) };
	ListNode<Edge<string>>* user1RelationUser2{ friends.findEdge(user1, user2) };

	if (user1Node == nullptr || user1RelationUser2 == nullptr || user1RelationUser2->data.weight != "pending") {
		return false;
	}

	return true;
}

bool Friend::haveNoRelation(const string& user1, const string& user2) {
	ListNode<GraphNode<string>>* user1Node{ friends.findNode(user1) };
	ListNode<Edge<string>>* user1RelationUser2{ friends.findEdge(user1, user2) };

	if (user1Node == nullptr || user1RelationUser2 == nullptr) {
		return true;
	}

	return false;
}

void Friend::addUser(const string& name) {
	friends.addNode(name);
}

void Friend::removeUser(const string& user) {
	friends.removeNode(user);
}

void Friend::sendFriendRequest(const string& from, const string& to) {
	friends.addEdge(from, to, "pending");
}

void Friend::acceptFriendRequest(const string& from, const string& to) {
	friends.changeEdgeWeight(from, to, "active");
}

void Friend::blockActiveFriend(const string& from, const string& to) {
	friends.removeEdge(from, to);
}

Graph<string> Friend::getFriends() const {
	return friends;
}

List<Edge<string>> Friend::getFriendList(const string& name) {
	return friends.findNode(name)->data.adjacencyList;
}

void Friend::viewFriends(const string& user) {
	ListNode<Edge<string>>* current{ friends.findNode(user)->data.adjacencyList.getHead() };

	if (current != nullptr) {
		cout << "<-------------------->\n";
		while (current != nullptr) {
			cout << "Name: " << current->data.data << ", Status: " << current->data.weight << endl;
			cout << "<-------------------->\n";
			current = current->next;
		}
	}
	else {
		cout << "<-------------------->\n";
		cout << user << " have no any friend.\n";
		cout << "<-------------------->\n";
	}
}

File: Mini-Instagram-Python/Friend.py
from Graph import Graph
from Tree import Tree


class Friend:
    def __init__(self):
        self.friends = Graph()

    def haveActiveRelation(self, user1, user2):
        user1Node = self.friends.findNode(user1)
        user1RelationUser2 = self.friends.findEdge(user1, user2)
        if user1Node and user1RelationUser2 and user1RelationUser2.weight == "active":
            return True
        return False
    
    def havePendingRelation(self, user1, user2):
        user1Node = self.friends.findNode(user1)
        user1RelationUser2 = self.friends.findEdge(user1, user2)
        if user1Node and user1RelationUser2 and user1RelationUser2.weight == "pending":
            return True
        return False
    
    def haveNoRelation(self, user1, user2):
        user1Node = self.friends.findNode(user1)
        user1RelationUser2 = self.friends.findEdge(user1, user2)
        if user1Node == None or user1RelationUser2 == None:
            return True
        return False
    
    def addUser(self, user):
        self.friends.addNode(user)
    
    def removeUser(self, user):
        self.friends.removeNode(user)

    def sendFriendRequest(self, fromNode, toNode):
        self.friends.addEdge(fromNode, toNode, "pending")
    
    def acceptFriendRequest(self, fromNode, toNode):
        self.friends.changeEdgeWeight(fromNode, toNode, "active")

    def blockActiveFriend(self, fromNode, toNode):
        self.friends.removeEdge(fromNode, toNode)

    def sugestFriends(self, user):
        userFriendList = self.getFriendList(user)
        userFriends = userFriendList.getHead()
        suggestions = Tree()
        while userFriends:
            friendFriendList = self.getFriendList(userFriends.data.data)
            friendFriends = friendFriendList.getHead()
            while friendFriends:
                if user != friendFriends.data.data and self.haveNoRelation(user, friendFriends.data.data):
                    suggestions.insert(friendFriends.data.data)
                friendFriends = friendFriends.next
            userFriends = userFriends.next
        return suggestions

    def getFriends(self):
        return self.friends

    def getFriendList(self, user):
        return self.friends.findNode(user).data.adjacencyList

    def viewFriends(self, user):
        current = self.findNode(user).data.adjacencyList.getHead()
        if current:
            print("<-------------------->")
            while current:
                print(f"Name: {current.data.data}, Status: {current.data.weight}")
                print("<-------------------->")
                current = current.next
        else:
            print("<-------------------->")
            print(f"{user} has no any friend.")
            print("<-------------------->")